python_modules/dagster/dagster/core/log_manager.py

```diff
@@ -0,0 +1,246 @@
+import datetime
+import itertools
+import logging
+import uuid
+
+from dagster import check, seven
+
+
+DAGSTER_META_KEY = 'dagster_meta'
+
+
+def _dump_value(value):
+    # dump namedtuples as objects instead of arrays
+    if isinstance(value, tuple) and hasattr(value, '_asdict'):
+        return seven.json.dumps(value._asdict())
+
+    return seven.json.dumps(value)
+
+
+def _kv_message(all_items):
+    sep = '\n'
+    format_str = '{key:>20} = {value}'
+    return sep + sep.join(
+        [format_str.format(key=key, value=_dump_value(value)) for key, value in all_items]
+    )
+
+
+class DagsterLogManager:
+    '''Centralized dispatch for logging through the execution context.
+
+    Handles the construction of uniform structured log messages and passes through to the underlying
+    loggers.
+
+    An instance of the log manager is made available to solids as context.log.
+
+    In an attempt to support the range of Python logging possibilities, the log manager can be
+    invoked in three ways:
+
+        1. Using standard convenience methods like those built in to the Python logging library:
+           context.log.{debug, info, warning, error, critical}
+        2. With any user-defined log level, like context.log.trace, or context.log.notice. Users are
+           expected to register these levels with the Python logging library (using
+           logging.addLevelName) when instantiating custom loggers.
+        3. Using the underlying integer API directly by calling, e.g., context.log.log(5, msg).
+
+    Args:
+        run_id (str): The run_id.
+        tags (dict): Tags for the run
+        loggers (List[logging.Logger]): Loggers to invoke.
+    '''
+
+    def __init__(self, run_id, logging_tags, loggers):
+        self.run_id = check.str_param(run_id, 'run_id')
+        self.logging_tags = check.dict_param(logging_tags, 'logging_tags')
+        self.loggers = check.list_param(loggers, 'loggers', of_type=logging.Logger)
+
+    def _log(self, level, orig_message, message_props):
+        '''Actually invoke the underlying loggers.
+
+        Args:
+            level (Union[str, int]): A built=in logging level name, a level name as registered using
```

Comment:
so is this just because the underlying logging systems support both variants? Would be really nice just to stick to one

Reply:
Yeah, we can simplify this implementation so that `DagsterLogManager.log()` takes only ints, and then the underlying `_log()` just uses ints internally

Reply:
This is addressed in 17956e

Reply:
The caveat is if someone has an implementation like
```
def debug(...):
    do stuff
    self.log(10, msg, ...)
```